### PR TITLE
Global meter forwarding implementation

### LINKED ISSUE #388: Make the global `Meter` useful for registering metric instruments before the SDK is initialized
There is a stale OTEP here: https://github.com/open-telemetry/oteps/pull/45
It may not describe a real problem in many languages, but it makes a proposal that should at least be followed in Go and probably several other languages.

To make concrete progress in this repository, we should make the global `Meter` instance a new "forwarding" implementation.  It will support allocating metric instruments and label sets before the global instance is registered.  It can use a `sync.Once` object for this purpose. Once the global instance is registered, objects returned by the global instance should begin forwarding to the installed SDK.

Otherwise, metrics allocated on the global Meter implementation are no-op instruments that will remain no-op instruments for the process lifetime.

### PR DESCRIPTION
**Pull request by jmacd**
Implements deferred initialization for metric instruments registered before the first Meter SDK is installed.

This implements the behavior specified in https://github.com/open-telemetry/oteps/pull/45.
Resolves #388.

### CODE REVIEW

File: internal/metric/mock.go [hunk 1]
```diff
@@ -71,13 +77,19 @@ const (
 )
 
 func (i *Instrument) AcquireHandle(labels apimetric.LabelSet) apimetric.HandleImpl {
+	if ld, ok := labels.(apimetric.LabelSetDelegate); ok {
+		labels = ld.Delegate()
+	}
```
Comment (rghetia) [inline internal/metric/mock.go]: This check should be done for sdk as well.

File: api/global/internal/meter_test.go [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+
+	require.Equal(t, map[core.Key]core.Value{
+		lvals1.Key: lvals1.Value,
+	}, mock.MeasurementBatches[2].LabelSet.Labels)
+	require.Equal(t, 1, len(mock.MeasurementBatches[2].Measurements))
+	require.Equal(t, core.NewFloat64Number(3),
+		mock.MeasurementBatches[2].Measurements[0].Number)
+	require.Equal(t, "test.measure",
+		mock.MeasurementBatches[2].Measurements[0].Instrument.Name)
+}
```
Comment (rghetia) [inline api/global/internal/meter_test.go]: Perhaps add test for multiple Meters.

File: internal/metric/mock.go (hunk 1 above)
Comment (jmacd) [reply internal/metric/mock.go]: Yes! This crossed my mind as I was walking out the office. Suggests more tests, too :)

**Comment by jmacd** [comment]
FYI I discovered that named Tracers will need similar support as discussed in https://github.com/open-telemetry/oteps/pull/74 but I will leave that out of this PR.

File: api/global/global.go [hunk 3]
```diff
@@ -15,55 +15,36 @@
...
 func SetTraceProvider(tp trace.Provider) {
-	globalTracer.Store(traceProvider{tp: tp})
+	internal.SetTraceProvider(tp)
 }
 
-// MeterProvider returns the registered global meter provider.
-// If none is registered then an instance of metric.NoopProvider is returned.
-// Use the trace provider to create a named meter. E.g.
+// MeterProvider returns the registered global meter provider.  If
+// none is registered then a dewfault meter provider is returned that
```
[the remark this answers is not in the document]
Comment (jmacd) [reply api/global/global.go]: Done.

File: api/global/internal/meter_test.go (hunk 2 above)
Comment (jmacd) [reply api/global/internal/meter_test.go]: Done.

File: api/global/internal/state.go [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+
+// SetMeterProvider is the internal implementation for global.SetMeterProvider.
+func SetMeterProvider(mp metric.Provider) {
+	delegateMeterOnce.Do(func() {
+		current := MeterProvider()
+
+		if current == mp {
+			// Setting the provider to the prior default
+			// is nonsense, set it to a noop.
+			mp = metric.NoopProvider{}
```
[the remark this answers is not in the document]
Comment (jmacd) [reply api/global/internal/state.go]: Done.

File: internal/metric/mock.go (hunk 1 above)
Comment (jmacd) [reply internal/metric/mock.go]: Done.

File: api/global/internal/meter.go [hunk 5]
```diff
@@ -0,0 +1,293 @@
...
+// The implementation here uses Mutexes to maintain a list of active
+// Meters in the Provider and Instruments in each Meter, under the
+// assumption that these interfaces are not performance-critical.
+//
+// We have the invariant that setDelegate() will be called before a
+// new metric.Provider implementation is registered as the global
+// provider.  Mutexes in the Provider and Meters ensure that each
+// instrument has a delegate before the global provider is set.
+//
+// LabelSets are implemented using the by delegating to the Meter
```
[the remark this answers is not in the document]
Comment (jmacd) [reply api/global/internal/meter.go]: Done.

File: api/global/internal/meter.go [hunk 6]
```diff
@@ -0,0 +1,293 @@
...
+// instance using the metric.LabelSetDelegator interface.
+//
+// Bound instrument operations are implementing by delegating to the
```
[the remark this answers is not in the document]
Comment (jmacd) [reply api/global/internal/meter.go]: Done. Thanks!

**Pull request merged**